### PR TITLE
Patch/insert payment fix

### PR DESCRIPTION
**Pull request by toneo-**
Patched the build error in InsertPaymentService. Also refactored it to use our injected repos/services consistently with the rest of the project.

@JoeDicker - You know how this part worked better than any of us - can you check to see that I didn't break anything even more? I also can't find a way to actually submit a payment through the UI

### CODE REVIEW

**Comment by lenfred123** [comment]
Ah - I've added a submit claim button in my PR too. I'll remove mine.

**Comment by toneo-** [comment]
@lenfred123 - submit claim button was added by Joe!

File: XYZDrivers/src/main/java/com/xyzdrivers/controllers/AuthController.java [hunk 1]
```diff
@@ -68,25 +78,26 @@ protected void doPost(HttpServletRequest request, HttpServletResponse response)
         
         User loggedInUser = null;
         HttpSession session = request.getSession(true);
-                
-        try {
-            if (userService.checkLoginDetails(username, password))
-            {
```
Comment (bkneis) [inline XYZDrivers/src/main/java/com/xyzdrivers/controllers/AuthController.java:74]: same line please

File: XYZDrivers/src/main/java/com/xyzdrivers/controllers/AuthController.java [hunk 2]
```diff
@@ -68,25 +78,26 @@ protected void doPost(HttpServletRequest request, HttpServletResponse response)
...
-                session.setAttribute("username", username);
-                loggedInUser = userRepo.get(username);
+        
+        if (username != null && password != null)
+        {      
+            try {            
+                if (userService.checkLoginDetails(username, password))
+                {
```
Comment (bkneis) [inline XYZDrivers/src/main/java/com/xyzdrivers/controllers/AuthController.java]: and here and one above ;)

File: XYZDrivers/src/main/java/com/xyzdrivers/controllers/PaymentController.java [hunk 3]
```diff
@@ -45,25 +48,34 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response)
...
+            amount = Float.parseFloat(request.getParameter("amount"));
+        }
+        catch (NumberFormatException ex)
+        {
+            amount = 0;
+        }
 
-            InsertPaymentService ips = new InsertPaymentService(con);
+        if (amount <= 0)
+        {
```
Comment (bkneis) [inline XYZDrivers/src/main/java/com/xyzdrivers/controllers/PaymentController.java]: same line

File: XYZDrivers/src/main/java/com/xyzdrivers/models/Member.java [hunk 4]
```diff
@@ -12,44 +12,77 @@
...
+    public LocalDate getDateOfBirth() {
+        return dateOfBirth;
+    }
+
+    public void setDateOfBirth(LocalDate dob) {
+        this.dateOfBirth = dob;
+    }
+
+    public LocalDate getDateOfRegistration() {
+        return dateOfRegistration;
```
Comment (bkneis) [inline XYZDrivers/src/main/java/com/xyzdrivers/models/Member.java:75]: Can we add here a sanity check to ensure the date is not in the future?

File: XYZDrivers/src/main/java/com/xyzdrivers/models/Member.java [hunk 5]
```diff
@@ -12,44 +12,77 @@
...
+    public void setDateOfRegistration(LocalDate dor) {
+        this.dateOfRegistration = dor;
     }
+
     public String getStatus() {
         return status;
     }
+
+    public void setStatus(String status) {
+        this.status = status;
```
Comment (bkneis) [inline XYZDrivers/src/main/java/com/xyzdrivers/models/Member.java]: Sanity check to ensure member status is in list of allowed statuses would be nice here too.

File: XYZDrivers/src/main/java/com/xyzdrivers/models/Member.java [hunk 6]
```diff
@@ -12,44 +12,77 @@
...
+    }
+
     public double getBalance() {
         return balance;
     }
+
+    public void setBalance(double balance) {
+        this.balance = balance;
```
Comment (bkneis) [inline XYZDrivers/src/main/java/com/xyzdrivers/models/Member.java]: ensure its positive?

File: XYZDrivers/src/main/java/com/xyzdrivers/services/InsertPaymentService.java [hunk 7]
```diff
@@ -1,101 +1,60 @@
...
+        if (mp.getPaymentAmount() <= 0)
+        {
+            throw new IllegalArgumentException("Given MembershipPayment must have a payment amount >= 0.");
         }
-
-        balance = currentBalance - mp.getPaymentAmount();
         
-        String query3 = "UPDATE MEMBERS SET BALANCE = ? WHERE ID = ?";
+        if (mp.getPaymentType() == null)
+        {
```
Comment (bkneis) [inline XYZDrivers/src/main/java/com/xyzdrivers/services/InsertPaymentService.java]: You love a good sanity check ;) nice

File: XYZDrivers/src/main/webapp/error.jsp [hunk 8]
```diff
@@ -1,17 +1,18 @@
...
 <html>
     <head>
         <title>Fatal Error</title>
         <meta charset="UTF-8">
         <meta name="viewport" content="width=device-width, initial-scale=1.0">
-        <link rel="stylesheet" type="text/css" href="node_modules/bootstrap/dist/css/bootstrap.min.css" />
-        <link rel="stylesheet" type="text/css" href="styles/main.css" />
     </head>
     <body>
-        <jsp:include page="nav.jsp"></jsp:include>
```
Comment (bkneis) [inline XYZDrivers/src/main/webapp/error.jsp]: Why remove the nav? Would make it easier for the user to get back to where they were?

**Review by bkneis — COMMENTED** [review]
Great work, just 4 - 5 small things

**Review by bkneis — CHANGES_REQUESTED** [review]
accidentally commented the last one

File: XYZDrivers/src/main/java/com/xyzdrivers/models/Member.java (hunk 4 above)
Comment (toneo-) [reply XYZDrivers/src/main/java/com/xyzdrivers/models/Member.java:75]: Good idea, ~~but not here - it should be checked when the member is created, more of a Repo thing?
If agreed I'll chuck that in this PR too!~~

Actually we can put the sanity check on setDateOfRegistration() if you like!

File: XYZDrivers/src/main/webapp/error.jsp (hunk 8 above)
Comment (toneo-) [reply XYZDrivers/src/main/webapp/error.jsp]: Good catch, accidentally deleted!

**Comment by toneo-** [comment]
@bkneis Review changes made

File: XYZDrivers/src/main/java/com/xyzdrivers/models/Member.java (hunk 4 above)
Comment (bkneis) [reply XYZDrivers/src/main/java/com/xyzdrivers/models/Member.java:75]: yeah give the mode some business logic

**Review by bkneis — APPROVED** [review]
LGTM

**Pull request merged**